### PR TITLE
Horizontal Pod Autoscaling e2e tests

### PR DESCRIPTION
**Pull request by socaa**
Part of #11570

@piosz 
@jszczepkowski

### CODE REVIEW

File: test/e2e/autoscaling_utils.go [hunk 1]
```diff
@@ -114,13 +120,30 @@ func (rc *ConsumingRC) sendOneConsumeCPUrequest(milicores int, durationSec int)
...
+	replicationController, err := rc.framework.Client.ReplicationControllers(rc.framework.Namespace.Name).Get(rc.name)
+	expectNoError(err)
+	if replicationController == nil {
+		Failf(rcIsNil)
+	}
+	return replicationController.Status.Replicas
+}
+
+func (rc *ConsumingRC) ReplicasEqual(desiredReplicas int) {
+	Logf("REPLICAS: %d", rc.GetReplicas())
```
Comment (jszczepkowski) [inline test/e2e/autoscaling_utils.go]: Make this log more readable.

File: test/e2e/horizontal_pod_autoscaling.go [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+	BeforeEach(func() {
+	})
+
+	AfterEach(func() {
+	})
+	It("should scale from 1 pod to 3 pods (scale resource: CPU)", func() {
+		rc = NewConsumingRC("rc", 1, "0.3", f)
+		rc.ConsumeCPU(700)
+		time.Sleep(sleep)
```
Comment (jszczepkowski) [inline test/e2e/horizontal_pod_autoscaling.go]: I think this test is waiting a lot of time (it waits always 10 minutes here). I will be better to wait for no more than 10 minutes and periodically query the number of replicas. See function waitForGroupSize (resize_nodes.go:95).

File: test/e2e/autoscaling_utils.go [hunk 3]
```diff
@@ -147,3 +170,30 @@ func startService(c *client.Client, ns, name string, replicas int) {
 	}
 	expectNoError(RunRC(config))
 }
+
+func createHorizontalPodAutoscaler(name, ns, cpu string) *client.ExperimentalClient {
+	config, err := loadConfig()
+	expectNoError(err)
+	expClient, errExpClient := client.NewExperimental(config)
+	expectNoError(errExpClient)
```
Comment (jszczepkowski) [inline test/e2e/autoscaling_utils.go]: After recent changes ExperimentalClient is already available as a part of regular Client. Just call rc.framework.Client.Experimental().

File: test/e2e/autoscaling_utils.go [hunk 4]
```diff
@@ -50,16 +54,18 @@ type ConsumingRC struct {
 	framework *Framework
 	channel   chan int
 	stop      chan int
+	expClient *client.ExperimentalClient
 }
 
 // NewConsumingRC creates new ConsumingRC
-func NewConsumingRC(name string, replicas int, framework *Framework) *ConsumingRC {
-	startService(framework.Client, framework.Namespace.Name, name, replicas)
+func NewConsumingRC(name string, replicas int, cpu string, framework *Framework) *ConsumingRC {
```
Comment (jszczepkowski) [inline test/e2e/autoscaling_utils.go]: I think we shouldn't mix ConsumingRC with HorizontalPodAutoscaler. Method createHorizontalPodAutoscaler should be moved to horizontal_pod_autoscaling.go and called directly from test. Argument cpu in this method should define initial cpu consumption of the rc.

File: test/e2e/autoscaling_utils.go [hunk 5]
```diff
@@ -114,13 +120,30 @@ func (rc *ConsumingRC) sendOneConsumeCPUrequest(milicores int, durationSec int)
...
 func (rc *ConsumingRC) CleanUp() {
 	rc.stop <- 0
 	expectNoError(DeleteRC(rc.framework.Client, rc.framework.Namespace.Name, rc.name))
 	expectNoError(rc.framework.Client.Services(rc.framework.Namespace.Name).Delete(rc.name))
+	expectNoError(rc.expClient.HorizontalPodAutoscalers(rc.framework.Namespace.Name).Delete(rc.name, api.NewDeleteOptions(0)))
 }
 
-func startService(c *client.Client, ns, name string, replicas int) {
+func startServiceAndRunRC(c *client.Client, ns, name string, replicas int) {
 	c.Services(ns).Create(&api.Service{
```
Comment (jszczepkowski) [inline test/e2e/autoscaling_utils.go]: Add error checking.

File: test/e2e/autoscaling_utils.go [hunk 6]
```diff
@@ -50,16 +54,18 @@ type ConsumingRC struct {
 	framework *Framework
 	channel   chan int
 	stop      chan int
+	expClient *client.ExperimentalClient
```
Comment (jszczepkowski) [inline test/e2e/autoscaling_utils.go]: Not needed (can be take from framework).

File: test/e2e/autoscaling_utils.go [hunk 7]
```diff
@@ -50,16 +54,18 @@ type ConsumingRC struct {
 	framework *Framework
```
Comment (jszczepkowski) [inline test/e2e/autoscaling_utils.go]: I would rename it: ConsumingRC -> ResourceConsumer.

File: test/e2e/autoscaling_utils.go [hunk 8]
```diff
@@ -114,13 +120,30 @@ func (rc *ConsumingRC) sendOneConsumeCPUrequest(milicores int, durationSec int)
...
+
 func (rc *ConsumingRC) CleanUp() {
 	rc.stop <- 0
 	expectNoError(DeleteRC(rc.framework.Client, rc.framework.Namespace.Name, rc.name))
 	expectNoError(rc.framework.Client.Services(rc.framework.Namespace.Name).Delete(rc.name))
+	expectNoError(rc.expClient.HorizontalPodAutoscalers(rc.framework.Namespace.Name).Delete(rc.name, api.NewDeleteOptions(0)))
 }
 
-func startService(c *client.Client, ns, name string, replicas int) {
+func startServiceAndRunRC(c *client.Client, ns, name string, replicas int) {
```
Comment (jszczepkowski) [inline test/e2e/autoscaling_utils.go]: I think better name will be: runServiceAndRCForResourceConsumer

File: test/e2e/horizontal_pod_autoscaling.go [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+var _ = Describe("Horizontal pod autoscaling", func() {
+	var rc *ConsumingRC
+	f := NewFramework("horizontal-pod-autoscaling")
+
+	BeforeEach(func() {
+	})
+
+	AfterEach(func() {
+	})
+	It("should scale from 1 pod to 3 pods (scale resource: CPU)", func() {
```
Comment (jszczepkowski) [inline test/e2e/horizontal_pod_autoscaling.go]: Please mark all test cases as skipped (we don't want them to start executing yet).

Comment (piosz) [inline test/e2e/horizontal_pod_autoscaling.go]: see here for reference https://github.com/kubernetes/kubernetes/blob/master/test/e2e/autoscaling.go#L66

File: test/e2e/autoscaling_utils.go (hunk 7 above)
Comment (socaa) [inline test/e2e/autoscaling_utils.go]: done

File: test/e2e/autoscaling_utils.go (hunk 1 above)
Comment (socaa) [inline test/e2e/autoscaling_utils.go]: done

File: test/e2e/horizontal_pod_autoscaling.go (hunk 2 above)
Comment (socaa) [inline test/e2e/horizontal_pod_autoscaling.go]: done

File: test/e2e/autoscaling_utils.go (hunk 3 above)
Comment (socaa) [inline test/e2e/autoscaling_utils.go]: done

File: test/e2e/autoscaling_utils.go (hunk 4 above)
Comment (socaa) [inline test/e2e/autoscaling_utils.go]: done

File: test/e2e/autoscaling_utils.go (hunk 5 above)
Comment (socaa) [inline test/e2e/autoscaling_utils.go]: done

File: test/e2e/autoscaling_utils.go (hunk 6 above)
Comment (socaa) [inline test/e2e/autoscaling_utils.go]: done

File: test/e2e/autoscaling_utils.go (hunk 8 above)
Comment (socaa) [inline test/e2e/autoscaling_utils.go]: done

File: test/e2e/horizontal_pod_autoscaling.go (hunk 9 above)
Comment (socaa) [inline test/e2e/horizontal_pod_autoscaling.go]: done

**Comment by socaa** [comment]
PTAL

@piosz 
@jszczepkowski

File: test/e2e/autoscaling_utils.go [hunk 10]
```diff
@@ -114,14 +115,37 @@ func (rc *ConsumingRC) sendOneConsumeCPUrequest(milicores int, durationSec int)
...
+func (rc *ResourceConsumer) GetReplicas() int {
+	replicationController, err := rc.framework.Client.ReplicationControllers(rc.framework.Namespace.Name).Get(rc.name)
+	expectNoError(err)
+	if replicationController == nil {
+		Failf(rcIsNil)
+	}
+	return replicationController.Status.Replicas
+}
+
+func (rc *ResourceConsumer) ReplicasEqual(desiredReplicas int) {
```
Comment (jszczepkowski) [inline test/e2e/autoscaling_utils.go]: I would call this WaitForReplicas

File: test/e2e/autoscaling_utils.go [hunk 11]
```diff
@@ -114,14 +115,37 @@ func (rc *ConsumingRC) sendOneConsumeCPUrequest(milicores int, durationSec int)
...
+	timeout := 10 * time.Minute
+	for start := time.Now(); time.Since(start) < timeout; time.Sleep(5 * time.Second) {
```
Comment (jszczepkowski) [inline test/e2e/autoscaling_utils.go]: Querying every 5 seconds is probably too often (the whole timeout is 10 minutes). I suggest increasing this to 20 seconds.

File: test/e2e/autoscaling_utils.go [hunk 12]
```diff
@@ -35,43 +35,44 @@ const (
...
 	framework *Framework
 	channel   chan int
 	stop      chan int
 }
 
-// NewConsumingRC creates new ConsumingRC
-func NewConsumingRC(name string, replicas int, framework *Framework) *ConsumingRC {
-	startService(framework.Client, framework.Namespace.Name, name, replicas)
-	rc := &ConsumingRC{
+// NewResourceConsumer creates new ResourceConsumer
```
Comment (jszczepkowski) [inline test/e2e/autoscaling_utils.go]: Pleas comment what is the unit of cpu argument (milicores?)

**Comment by jszczepkowski** [comment]
The tests seems to be broken:

hack/verify-boilerplate.sh
Boilerplate header is wrong for: /home/travis/gopath/src/github.com/kubernetes/kubernetes/test/e2e/horizontal_pod_autoscaling.go

File: test/e2e/autoscaling_utils.go (hunk 10 above)
Comment (socaa) [inline test/e2e/autoscaling_utils.go]: done

File: test/e2e/autoscaling_utils.go (hunk 11 above)
Comment (socaa) [inline test/e2e/autoscaling_utils.go]: done

File: test/e2e/autoscaling_utils.go (hunk 12 above)
Comment (socaa) [inline test/e2e/autoscaling_utils.go]: done

**Comment by socaa** [comment]
PTAL

@piosz 
@jszczepkowski

**Comment by jszczepkowski** [comment]
LGTM
Great thanks!

File: test/e2e/autoscaling_utils.go [hunk 13]
```diff
@@ -114,14 +116,37 @@ func (rc *ConsumingRC) sendOneConsumeCPUrequest(milicores int, durationSec int)
...
+		}
+	}
+	Failf("timeout waiting %v for pods size to be %d", timeout, desiredReplicas)
+}
+
+func (rc *ResourceConsumer) CleanUp() {
 	rc.stop <- 0
 	expectNoError(DeleteRC(rc.framework.Client, rc.framework.Namespace.Name, rc.name))
 	expectNoError(rc.framework.Client.Services(rc.framework.Namespace.Name).Delete(rc.name))
+	expectNoError(rc.framework.Client.Experimental().HorizontalPodAutoscalers(rc.framework.Namespace.Name).Delete(rc.name, api.NewDeleteOptions(0)))
```
Comment (piosz) [inline test/e2e/autoscaling_utils.go]: @socaa This is a very bad pattern when you clean up here resources created by someone else. Please move this call to test/e2e/horizontal_pod_autoscaling.go

Comment (socaa) [inline test/e2e/autoscaling_utils.go]: Horizontal Pod Autoscaler is deleted along with namespace (changes in new PR: #13786 )

**Comment by jszczepkowski** [comment]
Part of #12087.

**Pull request merged**